packages/@aws-cdk/example-construct-library/lib/example-resource.ts

```diff
@@ -0,0 +1,501 @@
+ * You do it with the '@resource' annotation:
+ *
+ * @resource AWS::CloudFormation::WaitCondition
+ */
+export class ExampleResource extends ExampleResourceBase {
+  /**
+   * Reference an existing ExampleResource,
+   * defined outside of the CDK code, by name.
+   *
+   * There might be other methods for referencing an existing resource,
+   * like fromExampleResourceArn,
+   * or fromExampleResourceAttributes
+   * (the last one if you want the importing behavior to be more customizable).
+   */
+  public static fromExampleResourceName(scope: core.Construct, id: string, exampleResourceName: string): IExampleResource {
+    // Imports are almost always implemented as a module-private
+    // inline class in the method itself.
+    // We extend ExampleResourceBase to reuse all of the logic inside it.
+    class Import extends ExampleResourceBase {
+      // we don't have an associated Role in this case
+      public readonly role = undefined;
+      // for imported resources, you always use the UnknownPrincipal,
+      // which ignores all modifications
+      public readonly grantPrincipal = new iam.UnknownPrincipal({ resource: this });
+
+      public readonly exampleResourceName = exampleResourceName;
+      // Since we have the name, we have to generate the ARN,
+      // using the Stack.formatArn helper method from the core library.
+      // We have to know the ARN components of ExampleResource in a few places, so,
+      // to avoid duplication, extract that into a module-private function
+      public readonly exampleResourceArn = core.Stack.of(scope)
+        .formatArn(exampleResourceArnComponents(exampleResourceName));
+    }
+
+    return new Import(scope, id);
+  }
+
+  // implement all fields that are abstract in ExampleResourceBase
+  public readonly exampleResourceArn: string;
+  public readonly exampleResourceName: string;
+  // while we know 'role' will actually never be undefined in this class,
+  // JSII does not allow changing the optionality of a field
+  // when overriding it, so it has to be 'role?'
+  public readonly role?: iam.IRole;
+  public readonly grantPrincipal: iam.IPrincipal;
+
+  /**
+   * The constructor of a construct has always 3 elements:
+   * the parent Construct, the string identifier,
+   * locally unique within the scope of the parent,
+   * and a properties struct.
+   *
+   * If the properties only have optional properties, like in our case,
+   * make sure to add a default value of an empty object to the props argument.
+   */
+  constructor(scope: core.Construct, id: string, props: ExampleResourceProps = {}) {
+    // Call the constructor from Resource superclass,
+    // which attaches this construct to the construct tree.
+    super(scope, id, {
+      // You need to let the Resource superclass know which of your properties
+      // signifies the resource's physical name.
+      // If your resource doesn't have a physical name,
```

Comment:
Most people don't understand what a physical name is. Perhaps explain how to differentiate?

Reply:
I linked to [this part of our guide](https://docs.aws.amazon.com/cdk/latest/guide/resources.html#resources_physical_names), I hope that's enough explanation.